build(windows): add arm64 support

rbqvq opened the pull request:
## Description
This PR adds an arm64 build for Windows.

## Issues Fixed or Closed
- Resolves https://github.com/orgs/LizardByte/discussions/898
- Resolves https://github.com/orgs/LizardByte/discussions/688

## Type of Change
* [ ]  Bug fix (non-breaking change which fixes an issue)
* [x]  New feature (non-breaking change which adds functionality)
* [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [x]  Dependency update (updates to dependencies)
* [x]  Documentation update (changes to documentation)
* [x]  Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
* [x]  My code follows the style guidelines of this project
* [x]  I have performed a self-review of my own code
* [x]  I have commented my code, particularly in hard-to-understand areas
* [x]  I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

rbqvq committed:
fix(build): using minhook-detours for arm64

MinHook does not support Windows ARM64.

So sing minhook-detours for Windows ARM64.

Signed-off-by: ReenigneArcher <42013603+ReenigneArcher@users.noreply.github.com>

rbqvq committed:
ci(build/windows): add support for arm64

Add CI build for Windows ARM64.

Signed-off-by: Coia Prant <coiaprant@gmail.com>
Co-authored-by: ReenigneArcher <42013603+ReenigneArcher@users.noreply.github.com>

rbqvq committed:
docs(windows): add documents for arm64

Add windows arm64 documents.

Signed-off-by: Coia Prant <coiaprant@gmail.com>
Co-authored-by: ReenigneArcher <42013603+ReenigneArcher@users.noreply.github.com>

rbqvq committed:
fix(build): disable Steam Audio Driver for Windows ARM64

Steam Audio Driver not available on Windows ARM64

Signed-off-by: Coia Prant <coiaprant@gmail.com>

rbqvq commented:
## Known Issues

It crashed after moonlight connect

```
[2025-05-25 12:58:21.029]: Info: [libx264 @ 0000024346e9af20] kb/s:629.76
[2025-05-25 12:58:21.059]: Info:
Device Description : Qualcomm(R) Adreno(TM) 8cx Gen 3
Device Vendor ID   : 0x4D4F4351
Device Device ID   : 0x36333630
Device Video Mem   : 0 MiB
Device Sys Mem     : 1 MiB
Share Sys Mem      : 16081 MiB
Feature Level      : 0x0000B100
Capture size       : 2560x1440
Offset             : 0x0
Virtual Desktop    : 2560x1440
[2025-05-25 12:58:21.059]: Info: Active GPU has HAGS disabled
[2025-05-25 12:58:21.059]: Info: Using realtime GPU priority
[2025-05-25 12:58:21.059]: Info:
Colorspace         : DXGI_COLOR_SPACE_RGB_FULL_G22_NONE_P709
Bits Per Color     : 8
Red Primary        : [0.673828,0.317383]
Green Primary      : [0.280273,0.660156]
Blue Primary       : [0.142578,0.0449219]
White Point        : [0.313477,0.329102]
Min Luminance      : 0.0745 nits
Max Luminance      : 445.761 nits
Max Full Luminance : 445.761 nits
[2025-05-25 12:58:21.059]: Info: Desktop resolution [2560x1440]
[2025-05-25 12:58:21.060]: Info: Desktop format [DXGI_FORMAT_B8G8R8A8_UNORM]
[2025-05-25 12:58:21.060]: Info: Display refresh rate [119.987Hz]
[2025-05-25 12:58:21.060]: Info: Requested frame rate [60fps]
[2025-05-25 12:58:21.065]: Info:
[2025-05-25 12:58:21.065]: Info: // Ignore any errors mentioned above, they are not relevant. //
[2025-05-25 12:58:21.065]: Info:
[2025-05-25 12:58:21.065]: Info: Found H.264 encoder: libx264 [software]
[2025-05-25 12:58:21.065]: Info: Executing [Desktop]
[2025-05-25 12:58:21.182]: Info: New streaming session started [active sessions: 1]
libc++abi: terminating due to uncaught exception of type boost::wrapexcept<std::out_of_range>: config out of range
```

ReenigneArcher commented on docs/getting_started.md:
```diff
@@ -280,24 +280,36 @@ brew uninstall sunshine
 overflow menu. Different versions of Windows may provide slightly different steps for uninstall.
 
-#### Standalone (lite version)
+#### Standalone / Portable (lite version)
```
revert, this will change any existing URLs

ReenigneArcher commented on src/config.cpp:
The changes in this file don't make sense.

rbqvq replied:
it necessary for clangarm64

rbqvq replied:
There is no problem.

After changing this file, clangarm64 works correctly

ReenigneArcher replied:
I meant, why did none of the other files require this change?

ReenigneArcher replied:
There must be something else causing this. We use `std` in plenty of other files.

Lapintam replied:
Hey guys, thanks for opening up this PR, have you been able to get Sunshine working with ARM64/apple silicon M1? I'd be happy to take a look and contribute to this effort. thanks!

rbqvq replied:
> Hey guys, thanks for opening up this PR, have you been able to get Sunshine working with ARM64/apple silicon M1? I'd be happy to take a look and contribute to this effort. thanks!

idk, i think it works, it has darwin arm64 target.

edesonmercy123 replied:
that's a great resolution

ReenigneArcher commented on src/platform/windows/audio.cpp:
```diff
@@ -38,7 +38,9 @@ DEFINE_PROPERTYKEY(PKEY_DeviceInterface_FriendlyName, 0x026e516e, 0xb814, 0x414b
 
   constexpr auto SAMPLE_RATE = 48000;
+#ifdef STEAM_DRIVER_SUBDIR
   constexpr auto STEAM_AUDIO_DRIVER_PATH = L"%CommonProgramFiles(x86)%\\Steam\\drivers\\Windows10\\" STEAM_DRIVER_SUBDIR L"\\SteamStreamingSpeakers.inf";
```
Where is steam installed on windows arm? Is it even available on ARM?

rbqvq replied:
Is this driver embedded in steam or installed by default?
woa x64 emulation cannot emulate kernel space driver.
but if it user space driver, it can emulate. (or steam provide a native arm64 installer)

I haven't used this function. Can you give me something to test?

rbqvq replied:
thx, i will try it

rbqvq replied:
Even steam virtual audio not support woa, but Virtual Audio Cable support it since pack45 :D

ReenigneArcher replied:
Alright, I don't think it will be necessary to have this macro though. Sunshine just won't do anything if the driver doesn't exist.

rbqvq replied:
For aarch64, STEAM_DRIVER_SUBDIR will not be defined, and STEAM_AUDIO_DRIVER_PATH will failed.

That way I add a check.

If not add it, we should define STEAM_DRIVER_SUBDIR on else branch.

ReenigneArcher commented on docs/building.md:
````diff
@@ -72,36 +72,73 @@ sudo port install "${dependencies[@]}"
 pacman -S "${dependencies[@]}"
 ```
 
+ℹ️ [UCRT64] Latest version of GCC is not compatible (Required)
+
+```bash
+wget https://repo.msys2.org/mingw/${MSYSTEM}/mingw-w64-${TOOLCHAIN}-gcc-14.2.0-3-any.pkg.tar.zst
+wget https://repo.msys2.org/mingw/${MSYSTEM}/mingw-w64-${TOOLCHAIN}-gcc-libs-14.2.0-3-any.pkg.tar.zst
+pacman -U --noconfirm mingw-w64-${TOOLCHAIN}-gcc-14.2.0-3-any.pkg.tar.zst mingw-w64-${TOOLCHAIN}-gcc-libs-14.2.0-3-any.pkg.tar.zst
+```
+
+ℹ️ Latest version of doxygen is not compatible (Optional)
+
+```bash
+wget https://repo.msys2.org/mingw/${MSYSTEM}/mingw-w64-${TOOLCHAIN}-doxygen-1.12.0-3-any.pkg.tar.zst
+pacman -U --noconfirm mingw-w64-${TOOLCHAIN}-doxygen-1.12.0-3-any.pkg.tar.zst
+```
+
````
```suggestion
```

We don't log bugs in our docs.

ReenigneArcher commented on docs/building.md:
```diff
@@ -72,36 +72,73 @@ sudo port install "${dependencies[@]}"
-First you need to install [MSYS2](https://www.msys2.org), then startup "MSYS2 UCRT64" and execute the following
-commands.
+
+ℹ️ Cross-compilation is not supported on Windows. You must build on the target architecture.
```
This looks copied from build-deps... we don't use this style in the doxygen docs. See https://docs.lizardbyte.dev/projects/sunshine/master/md_third-party_2doxyconfig_2docs_2source__code.html

ReenigneArcher commented on cmake/dependencies/windows.cmake:
```diff
@@ -1,9 +1,11 @@
+    add_library(minhook::minhook STATIC IMPORTED)
+    set_property(TARGET minhook::minhook PROPERTY IMPORTED_LOCATION ${MINHOOK_LIBRARY})
+    target_include_directories(minhook::minhook INTERFACE ${MINHOOK_INCLUDE_DIR})
+endif()
```
```suggestion
endif()

```

ReenigneArcher commented on .github/workflows/CI.yml:
```diff
@@ -1068,4 +1114,4 @@ jobs:
           prerelease: true
           tag: ${{ needs.setup_release.outputs.release_tag }}
-          token: ${{ secrets.GH_BOT_TOKEN }}
+          token: ${{ secrets.GH_BOT_TOKEN }}
```
```suggestion
          token: ${{ secrets.GH_BOT_TOKEN }}

```

All files should end with a blank empty line.

ReenigneArcher commented on docs/getting_started.md:
```diff
@@ -280,24 +280,36 @@ brew uninstall sunshine
+   | Architecture          | Installer                                                                                                                                    |
+   |-----------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
+   | AMD64/x64 (Intel/AMD) | [Sunshine-Windows-AMD64-installer.exe](https://github.com/LizardByte/Sunshine/releases/latest/download/Sunshine-Windows-AMD64-installer.exe) |
+   | ARM64                 | The installer is not currently supported, please use the portable version                                                                    |
```
Please use the link as I had in my branch.

The installer will be available before this PR is merged.

ReenigneArcher commented:
I think MF encoder is already included with the FFmpeg build, so why not?

rbqvq commented:
Build failed since https://github.com/LizardByte/Sunshine/issues/3936 resolve

rbqvq committed:
fix(build): clang namespace resolution

Use ::std:: fully qualified namespace in template functions to avoid
  ambiguity with clang namespace resolution.

Signed-off-by: Coia Prant <coiaprant@gmail.com>
Co-authored-by: Ricky8955555 <rkmiao@duck.com>

cgutman commented on src/platform/windows/display_base.cpp:
```diff
@@ -12,7 +12,11 @@
+
+// conditional includes
+#if defined(__x86_64) || defined(__x86_64__) || defined(__amd64) || defined(__amd64__) || defined(_M_AMD64)
+  #include <MinHook.h>
```
I don't think we can/should do this. It might work now by luck because none of the Windows ARM64 machines out in the wild have a discrete GPU, but it is definitely not guaranteed to continue working. We should look into hooking libraries that support ARM64, like [Detours](https://github.com/microsoft/Detours).

ReenigneArcher replied:
This should work: https://github.com/LizardByte/Sunshine/pull/4661/commits/7d2aa63471d09a172125956354f9cf09619f4d77

rbqvq committed:
fix(build): clang/llvm icon path error

Add TOSTRING for ICON.

Signed-off-by: Coia Prant <coiaprant@gmail.com>

rbqvq commented:
https://github.com/LizardByte/Sunshine/pull/4048 will cause llvm-rc fails
```
FILEVERSION fields (194208) does not fit in 16 bits.```

faratech committed:
feat: add Media Foundation encoder for Windows ARM64

Add Windows Media Foundation hardware encoder support for platforms
that don't have NVENC, AMF, or QSV (primarily Qualcomm Snapdragon).

Encoders added:
- h264_mf: H.264 via Media Foundation
- hevc_mf: HEVC via Media Foundation
- av1_mf: AV1 via Media Foundation

Configuration:
- Uses D3D11VA hardware device for GPU texture input
- CBR rate control with display_remoting scenario for streaming
- Only supports SDR 4:2:0 8-bit (Qualcomm MF encoder limitation)
- Fixed GOP size of 120 frames since MF doesn't support on-demand IDR

The mediafoundation encoder is probed after quicksync and amdvce,
serving as a fallback for ARM64 Windows devices.

faratech committed:
feat: add Qualcomm Adreno GPU detection

Add vendor ID detection for Qualcomm Adreno GPUs:
- 0x4D4F4351: "QCOM" in ASCII (reversed as little-endian)
- 0x5143: Alternate Qualcomm vendor ID

When a Qualcomm GPU is detected, only Media Foundation encoders
(h264_mf, hevc_mf, av1_mf) are considered compatible, as Qualcomm
doesn't support NVENC, AMF, or QSV encoder APIs.

ReenigneArcher commented:
wix installer is working now, could you test it in combination with the arm build?

https://github.com/LizardByte/Sunshine/pull/3916

I still need to make a few tweaks to that PR though.

rbqvq committed:
fix(build): using @rollup/wasm-node

CLANGARM64 cannot load @rollup/rollup-win32-arm64-msvc native version.

Add npm overrides to force rollup to use the WASM version.

Add a CI step to dump npm debug logs (always run) to aid troubleshooting install failures.

Enable nvm_node for Debian 12 and 13 in scripts/linux_build.sh to allow Node installation via nvm on those distros.

Signed-off-by: Coia Prant <coiaprant@gmail.com>
Co-authored-by: ReenigneArcher <42013603+ReenigneArcher@users.noreply.github.com>

ReenigneArcher committed:
chores(build): disable warnings for Windows ARM64

Delete the `#warning` emitted for non-x64 architectures in src/platform/windows/audio.cpp.

Update cmake/compile_definitions/windows.cmake to suppress specific GCC warnings when building for Windows ARM64.

Add flags to SUNSHINE_COMPILE_OPTIONS for CMAKE_SYSTEM_PROCESSOR=ARM64 to silence warnings.

- `-Wno-dll-attribute-on-redeclaration` for Boost
- `-Wno-unknown-warning-option` for ViGEmClient
- `-Wno-unused-variable` for Boost

Signed-off-by: ReenigneArcher <42013603+ReenigneArcher@users.noreply.github.com>

ReenigneArcher committed:
chores: use std::vector instead of VLAs on Windows

Replace variable-length arrays with std::vector in src/platform/windows/input.cpp and src/platform/windows/misc.cpp for safety and portability.

Add <vector> includes, allocate wide-character buffer with std::vector<WCHAR> and pass wide.data() to MultiByteToWideChar, and replace WSABUF VLA with std::vector<WSABUF> and use .data() for msg.lpBuffers.

These changes avoid non-standard VLAs and reduce stack-allocation risks.

Signed-off-by: ReenigneArcher <42013603+ReenigneArcher@users.noreply.github.com>

ReenigneArcher committed:
fix(build): drop unused HMODULE member from nvenc header

Drop the HMODULE dll = nullptr; declaration from src/nvenc/nvenc_d3d11_on_cuda.h.

This removes an unused module handle field from the nvenc D3D11-on-CUDA wrapper to reduce unnecessary state.

Signed-off-by: ReenigneArcher <42013603+ReenigneArcher@users.noreply.github.com>

ReenigneArcher commented on .github/workflows/ci-windows.yml:
```diff
@@ -169,17 +179,21 @@ jobs:
-          cpack -G NSIS
-          cpack -G ZIP
+          # nsis installer
+          if [[ ${CI_MSYSTEM} == "ucrt64" ]]; then
```
```suggestion
          if [[ ${MSYSTEM} == "ucrt64" ]]; then
```

ReenigneArcher commented on .github/workflows/ci-windows.yml:
```diff
@@ -169,17 +179,21 @@ jobs:
 
-          # move
-          mv ./cpack_artifacts/Sunshine.exe ../artifacts/Sunshine-${{ matrix.name }}-installer.exe
+          # portable version
```
TODO: add wix after https://github.com/LizardByte/Sunshine/pull/3916 is merged

ReenigneArcher commented on src/video.cpp:
```diff
@@ -825,6 +826,63 @@ namespace video {
+      AV_PIX_FMT_NV12,       // SDR 4:2:0 8-bit (only format Qualcomm supports)
+      AV_PIX_FMT_NONE,       // No HDR - Qualcomm MF only supports 8-bit
+      AV_PIX_FMT_NONE,       // No YUV444 SDR
+      AV_PIX_FMT_NONE,       // No YUV444 HDR
```
These comments are all formatted wrong. You can run clang format script in the scripts directory to autofix the code formatting.

rbqvq commented on .github/workflows/ci-windows.yml:
```diff
@@ -46,7 +51,9 @@ jobs:
-          MSYSTEM: ${{ matrix.msystem }}
+          # MSYSTEM is a built-in environment variable of MSYS2.
+          # Do not use this environment variable name.
+          MATRIX_MSYSTEM: ${{ matrix.msystem }}
```
Okay, Let me do it.

I don't think override the pre-defined env is right.

rbqvq commented on .github/workflows/ci-windows.yml:
```diff
@@ -169,17 +181,24 @@ jobs:
+          if [[ "${MATRIX_MSYSTEM}" == "ucrt64" ]]; then
+            cpack -G NSIS
+            mv ./cpack_artifacts/Sunshine.exe ../artifacts/Sunshine-${{ matrix.name }}-installer.exe
+          fi
```
Its works fine.
Done.

rbqvq commented:
@ReenigneArcher Thank you, I've rebase it

This is not my work alone, I need to thank everyone involved in this work

ReenigneArcher commented:
Thank you. Please no more force pushing.

ReenigneArcher committed:
docs(building): update syntax for github admonition

ReenigneArcher committed:
Add NSIS debug step to Windows CI

Insert a 'Debug nsis' step into .github/workflows/ci-windows.yml that always runs and prints the NSISOutput.log using the msys2 shell. This captures NSIS installer build logs for troubleshooting; the step uses '|| true' to avoid failing the job. The step is placed before the existing 'Debug wix' step.

ReenigneArcher committed:
Enable NSIS logging only on AMD64

Make NSIS detailed logging conditional on the target architecture by adding NSIS_LOGSET_COMMAND (set to "LogSet on" only when CMAKE_SYSTEM_PROCESSOR matches AMD64) and using it in CPACK_NSIS_EXTRA_INSTALL_COMMANDS and CPACK_NSIS_EXTRA_UNINSTALL_COMMANDS instead of always calling LogSet. Update docs/getting_started.md to note that the install.log under %PROGRAMFILES% is available on AMD64 only.

rbqvq commented:
@ReenigneArcher It should be fix in #4687

Actions for windows build still working:
https://github.com/rbqvq/Sunshine/actions/runs/21800375818

rbqvq commented:
- x86 

EdgeUpdate.exe

Tags: `peexe`, `signed`, `overlay`, `idle`
Magic: PE32 executable (GUI) Intel 80386, for MS Windows
Target Machine: Intel 386 or later processors and compatible processors

https://www.virustotal.com/gui/file/955b5950fd92e5c5eec395b022566c9a2594ca3126d894408e2e29ba25d8b0dd/details

- x64

bash.exe (Git For Windows)

Tags: `peexe`, `64bits`, `known-distributor`, `overlay`, `idle`
Magic: PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows
Target Machine: x64

https://www.virustotal.com/gui/file/fc362aae1f217d34d02237d99527839dd6cf85b37c7a0277109400bfa0faa594/details

- arm64

git.exe  (Git For Windows)

Tags: `peexe`, `arm`, `overlay`, `64bits`
Magic: PE32+ executable (console) Aarch64, for MS Windows
Target Machine: ARMv8 in 64-bit mode

https://www.virustotal.com/gui/file/bede8ae74caec9ae40c20910b360f01863042b5f69aa1c09fc2946dc9a52e100/details

---

- x64

dxgi-info.exe

Tags: `peexe`, `64bits`
Magic: PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows
Target Machine: x64

https://www.virustotal.com/gui/file/955b5950fd92e5c5eec395b022566c9a2594ca3126d894408e2e29ba25d8b0dd/details
